File: frontend/src/lib/actionItems.ts
// Frontend wrappers for the action-item, meeting-note and export commands.
// Keeps the Tauri command names and DTO shapes in one place — the Rust side
// serializes these structs with their field names as-is (snake_case), so the
// shapes below mirror `database::models::{ActionItem, MeetingNote}` exactly.

import { invoke } from "@tauri-apps/api/core";

export type ActionItemStatus = "open" | "done";

/** Where an item came from. `summary` items are owned by the extraction pass
 *  and get replaced when a summary is regenerated; `manual` and `agent` items
 *  are never touched by it. */
export type ActionItemSource = "summary" | "manual" | "agent";

export interface ActionItem {
  id: string;
  meeting_id: string;
  text: string;
  assignee: string | null;
  due_hint: string | null;
  status: ActionItemStatus;
  source: ActionItemSource;
  external_ref: string | null;
  created_at: string;
  updated_at: string;
  completed_at: string | null;
}

export interface MeetingNote {
  id: string;
  meeting_id: string;
  body: string;
  source: "manual" | "agent";
  created_at: string;
}

export type ExportFormat = "markdown" | "json";

export interface ExportResult {
  meeting_id: string;
  title: string;
  format: ExportFormat;
  /** Suggested filename, e.g. `weekly-sync-2026-07-15.md`. */
  filename: string;
  content: string;
}

/** Payload of the `action-items-extracted` event, emitted when the background
 *  extraction pass finishes for a meeting. */
export interface ActionItemsExtractedPayload {
  meeting_id: string;
  count: number;
}

export const ACTION_ITEMS_EXTRACTED_EVENT = "action-items-extracted";

/** Items for one meeting, or every item across every meeting when `meetingId`
 *  is omitted. */
export async function listActionItems(meetingId?: string): Promise<ActionItem[]> {
  return invoke<ActionItem[]>("list_action_items", {
    meetingId: meetingId ?? null,
  });
}

/** Open items across every meeting, newest first. */
export async function listOpenActionItems(): Promise<ActionItem[]> {
  return invoke<ActionItem[]>("list_open_action_items");
}

export async function createActionItem(
  meetingId: string,
  text: string,
  assignee?: string | null,
  dueHint?: string | null,
): Promise<ActionItem> {
  return invoke<ActionItem>("create_action_item", {
    meetingId,
    text,
    assignee: assignee ?? null,
    dueHint: dueHint ?? null,
  });
}

export async function setActionItemStatus(
  id: string,
  status: ActionItemStatus,
): Promise<ActionItem> {
  return invoke<ActionItem>("set_action_item_status", { id, status });
}

/** Patch an item. Omitted fields are left unchanged; passing `""` for
 *  `assignee`/`dueHint` clears them. */
export async function updateActionItem(
  id: string,
  fields: { text?: string; assignee?: string; dueHint?: string },
): Promise<ActionItem> {
  return invoke<ActionItem>("update_action_item", {
    id,
    text: fields.text ?? null,
    assignee: fields.assignee ?? null,
    dueHint: fields.dueHint ?? null,
  });
}

export async function deleteActionItem(id: string): Promise<boolean> {
  return invoke<boolean>("delete_action_item", { id });
}

/** Run extraction on demand for a meeting that already has a summary. Resolves
 *  to the number of items stored. Meetings summarized before this feature
 *  existed have no items until this runs (or the summary is regenerated). */
export async function extractActionItems(meetingId: string): Promise<number> {
  return invoke<number>("extract_action_items", { meetingId });
}

export async function listMeetingNotes(meetingId: string): Promise<MeetingNote[]> {
  return invoke<MeetingNote[]>("list_meeting_notes", { meetingId });
}

export async function addMeetingNote(
  meetingId: string,
  body: string,
): Promise<MeetingNote> {
  return invoke<MeetingNote>("add_meeting_note", { meetingId, body });
}

export async function deleteMeetingNote(id: string): Promise<boolean> {
  return invoke<boolean>("delete_meeting_note", { id });
}

/** Build a self-contained export document (summary + action items + full
 *  speaker-attributed transcript) without writing it anywhere. */
export async function exportMeeting(
  meetingId: string,
  format: ExportFormat,
): Promise<ExportResult> {
  return invoke<ExportResult>("export_meeting", { meetingId, format });
}

/** Export and prompt for a save location. Resolves to the written path, or
 *  `null` if the user cancelled the dialog. */
export async function exportMeetingToFile(
  meetingId: string,
  format: ExportFormat,
): Promise<string | null> {
  return invoke<string | null>("export_meeting_to_file", { meetingId, format });
}

File: frontend/src-tauri/src/api/action_items.rs
//! Tauri commands for action items and meeting notes.
//!
//! Thin wrappers over `ActionItemsRepository` / `MeetingNotesRepository`, plus
//! the on-demand extraction entry point. Items created through this surface are
//! `source = 'manual'` — the extractor owns `'summary'` rows and replaces them
//! on every re-run, so anything the user typed must be tagged differently to
//! survive a re-summarization.

use crate::database::models::{ActionItem, MeetingNote};
use crate::database::repositories::action_item::{
    ActionItemsRepository, NewActionItem, SOURCE_MANUAL,
};
use crate::database::repositories::meeting_note::{
    MeetingNotesRepository, SOURCE_MANUAL as NOTE_SOURCE_MANUAL,
};
use crate::database::repositories::setting::SettingsRepository;
use crate::database::repositories::summary::SummaryProcessesRepository;
use crate::state::AppState;
use crate::summary::action_extraction;
use crate::summary::transcript_action_items;
use crate::summary::markdown_export::extract_markdown;
use tauri::{AppHandle, Runtime};
use tracing::info;

/// Action items for one meeting (open first, then oldest-first within each
/// group), or every item across every meeting when `meeting_id` is omitted.
#[tauri::command]
pub async fn list_action_items<R: Runtime>(
    _app: AppHandle<R>,
    state: tauri::State<'_, AppState>,
    meeting_id: Option<String>,
) -> Result<Vec<ActionItem>, String> {
    let pool = state.db_manager.pool();
    match meeting_id {
        Some(mid) => ActionItemsRepository::list_by_meeting(pool, &mid)
            .await
            .map_err(|e| format!("Failed to list action items: {e}")),
        None => ActionItemsRepository::list_all(pool)
            .await
            .map_err(|e| format!("Failed to list action items: {e}")),
    }
}

/// Every open action item across all meetings, newest first.
#[tauri::command]
pub async fn list_open_action_items<R: Runtime>(
    _app: AppHandle<R>,
    state: tauri::State<'_, AppState>,
) -> Result<Vec<ActionItem>, String> {
    ActionItemsRepository::list_open(state.db_manager.pool(), None)
        .await
        .map_err(|e| format!("Failed to list open action items: {e}"))
}

#[tauri::command]
pub async fn create_action_item<R: Runtime>(
    _app: AppHandle<R>,
    state: tauri::State<'_, AppState>,
    meeting_id: String,
    text: String,
    assignee: Option<String>,
    due_hint: Option<String>,
) -> Result<ActionItem, String> {
    let text = text.trim().to_string();
    if text.is_empty() {
        return Err("Action item text cannot be empty".to_string());
    }

    let item = NewActionItem {
        text,
        assignee: assignee.and_then(non_empty),
        due_hint: due_hint.and_then(non_empty),
        ..Default::default()
    };

    ActionItemsRepository::create(state.db_manager.pool(), &meeting_id, &item, SOURCE_MANUAL)
        .await
        .map_err(|e| format!("Failed to create action item: {e}"))
}

/// Toggle an item between `"open"` and `"done"`.
#[tauri::command]
pub async fn set_action_item_status<R: Runtime>(
    _app: AppHandle<R>,
    state: tauri::State<'_, AppState>,
    id: String,
    status: String,
) -> Result<ActionItem, String> {
    ActionItemsRepository::set_status(state.db_manager.pool(), &id, &status)
        .await
        .map_err(|e| format!("Failed to update action item status: {e}"))?
        .ok_or_else(|| format!("Action item not found: {id}"))
}

/// Patch an item's editable fields. Omitted arguments are left unchanged;
/// passing an empty string for `assignee`/`due_hint` clears it.
#[tauri::command]
pub async fn update_action_item<R: Runtime>(
    _app: AppHandle<R>,
    state: tauri::State<'_, AppState>,
    id: String,
    text: Option<String>,
    assignee: Option<String>,
    due_hint: Option<String>,
) -> Result<ActionItem, String> {
    ActionItemsRepository::update(
        state.db_manager.pool(),
        &id,
        text.as_deref(),
        assignee.as_deref(),
        due_hint.as_deref(),
    )
    .await
    .map_err(|e| format!("Failed to update action item: {e}"))?
    .ok_or_else(|| format!("Action item not found: {id}"))
}

#[tauri::command]
pub async fn delete_action_item<R: Runtime>(
    _app: AppHandle<R>,
    state: tauri::State<'_, AppState>,
    id: String,
) -> Result<bool, String> {
    ActionItemsRepository::delete(state.db_manager.pool(), &id)
        .await
        .map_err(|e| format!("Failed to delete action item: {e}"))
}

/// Run action-item extraction for a meeting that already has a summary.
///
/// The automatic pass only fires when a summary finishes, so meetings
/// summarized before this feature shipped — or whose extraction failed at the
/// time — have no items until this is called. Unlike the automatic pass this
/// awaits the result and surfaces errors: the user asked for it and is watching
/// a spinner, so silence would be worse than a message.
#[tauri::command]
pub async fn extract_action_items<R: Runtime>(
    app: AppHandle<R>,
    state: tauri::State<'_, AppState>,
    meeting_id: String,
) -> Result<usize, String> {
    let pool = state.db_manager.pool();

    let config = SettingsRepository::get_model_config(pool)
        .await
        .map_err(|e| format!("Failed to read model config: {e}"))?
        .ok_or_else(|| "No summary model configured. Set one in settings first.".to_string())?;

    info!(
        "On-demand action-item extraction for {} using {}/{}",
        meeting_id, config.provider, config.model
    );

    // Prefer the transcript (timestamped ground truth). Fall back to the
    // summary only for meetings whose transcript segments aren't stored.
    match transcript_action_items::extract_from_transcript(
        &app,
        pool,
        &meeting_id,
        &config.provider,
        &config.model,
    )
    .await
    {
        Ok(count) => Ok(count),
        Err(transcript_err) => {
            info!(
                "Transcript extraction unavailable for {meeting_id} ({transcript_err}); \
                 falling back to the summary"
            );
            let process =
                SummaryProcessesRepository::get_summary_data_for_meeting(pool, &meeting_id)
                    .await
                    .map_err(|e| format!("Failed to load summary: {e}"))?
                    .ok_or_else(|| {
                        "No transcript to extract from, and no summary either. Generate a \
                         summary first."
                            .to_string()
                    })?;
            let result_str = process
                .result
                .ok_or_else(|| "This meeting has no summary yet. Generate one first.".to_string())?;
            let result_json: serde_json::Value = serde_json::from_str(&result_str)
                .map_err(|e| format!("Stored summary is not valid JSON: {e}"))?;
            let markdown = extract_markdown(&result_json)
                .filter(|m| !m.trim().is_empty())
                .ok_or_else(|| {
                    "Nothing to extract from: no transcript and an empty summary.".to_string()
                })?;

            action_extraction::extract_for_meeting(
                &app,
                pool,
                &meeting_id,
                &markdown,
                &config.provider,
                &config.model,
            )
            .await
        }
    }
}

/// Start live (in-meeting) action-item extraction for the current recording.
/// Gated by the frontend on the Beta feature flag; runs the summary model
/// periodically over the in-memory transcript and emits provisional items.
#[tauri::command]
pub async fn start_live_action_extraction<R: Runtime>(
    app: AppHandle<R>,
    state: tauri::State<'_, AppState>,
) -> Result<(), String> {
    let pool = state.db_manager.pool();
    let config = SettingsRepository::get_model_config(pool)
        .await
        .map_err(|e| format!("Failed to read model config: {e}"))?
        .ok_or_else(|| "No summary model configured. Set one in settings first.".to_string())?;

    crate::summary::live_action_items::start(app, pool.clone(), config.provider, config.model);
    Ok(())
}

#[tauri::command]
pub async fn stop_live_action_extraction() -> Result<(), String> {
    crate::summary::live_action_items::stop();
    Ok(())
}

#[tauri::command]
pub async fn add_meeting_note<R: Runtime>(
    _app: AppHandle<R>,
    state: tauri::State<'_, AppState>,
    meeting_id: String,
    body: String,
) -> Result<MeetingNote, String> {
    MeetingNotesRepository::create(
        state.db_manager.pool(),
        &meeting_id,
        &body,
        NOTE_SOURCE_MANUAL,
    )
    .await
    .map_err(|e| format!("Failed to add note: {e}"))
}

#[tauri::command]
pub async fn list_meeting_notes<R: Runtime>(
    _app: AppHandle<R>,
    state: tauri::State<'_, AppState>,
    meeting_id: String,
) -> Result<Vec<MeetingNote>, String> {
    MeetingNotesRepository::list_by_meeting(state.db_manager.pool(), &meeting_id)
        .await
        .map_err(|e| format!("Failed to list notes: {e}"))
}

#[tauri::command]
pub async fn delete_meeting_note<R: Runtime>(
    _app: AppHandle<R>,
    state: tauri::State<'_, AppState>,
    id: String,
) -> Result<bool, String> {
    MeetingNotesRepository::delete(state.db_manager.pool(), &id)
        .await
        .map_err(|e| format!("Failed to delete note: {e}"))
}

/// Trim, then drop empty strings. The frontend sends `""` for a cleared input;
/// storing that instead of NULL would render as an empty assignee chip.
fn non_empty(s: String) -> Option<String> {
    let t = s.trim();
    (!t.is_empty()).then(|| t.to_string())
}
